Change the root password spoke GUI design

VladimirSlavik opened the pull request:
This resolves ambiguity of what options are available when root is locked.

~~This is a draft because there are at least these TODOs:~~
- [x] make bold labels "clickable" 
- [x] verify the logic
- [x] find bugs etc. to link
- [x] consider renaming spoke to "Root account security"
- [x] ~~add an "empty password" banner~~ automatically switches to no password + locked instead.

See also: https://github.com/rhinstaller/anaconda/pull/3262#issuecomment-839202230
Suggested-by: Mairin Duffy <mairin@redhat.com> @mairin

VladimirSlavik commented:
@mairin, please comment on the preferred form of attribution, if you want it different.

VladimirSlavik commented:
Here's how it looks.
(That shall not change substantially, anymore.)

![Screenshot_vs_main_2021-07-14_16:24:05](https://user-images.githubusercontent.com/15903878/125640779-491e5ab9-44dd-4378-a905-3586e05352f8.png)

![Screenshot_vs_main_2021-07-14_16:24:12](https://user-images.githubusercontent.com/15903878/125640816-cd28cb07-2eb9-41c8-9076-2925907505f6.png)


https://user-images.githubusercontent.com/15903878/125641078-e552f276-c430-48a0-bdca-459347c0b3bf.mp4

mairin commented:
Wow that looks wonderful, @VladimirSlavik !!! The padding / margins / spacing etc. are right on and the animation looks slick. 🎉

poncovka commented on pyanaconda/ui/gui/spokes/root_password.py:
```diff
@@ -159,13 +160,16 @@ def refresh(self):
             self._users_module.RootPasswordSSHLoginAllowed
         )
-        if not self._lock.get_active():
+        if self._enable_radio.get_active():
```
Maybe `self._enable_radio.get_active()` could be hidden in a new property.

poncovka commented on pyanaconda/ui/gui/spokes/root_password.py:
```diff
@@ -179,7 +183,7 @@ def status(self):
             # reconfig mode currently allows re-enabling a locked root account if
             # user sets a new root password
-            if is_reconfiguration_mode() and not self._lock.get_active():
+            if is_reconfiguration_mode() and not self._enable_radio.get_active():
```
This needs to be tested with the initial setup.

poncovka reviewed (commented):
It looks great!

poncovka commented:
Please, test it with `rootpw --lock password`. I don't think we need to support this combination in UI, but the kickstart installation should still work. You can also run the `rootpw-lock` kickstart test.

jstodola commented:
It looks like a great improvement in usability of the spoke!

I would just like to express my dislike for the animation. Although it looks OK when the installation is running locally, animations in the installer in general do not look good when running a remote installation via VNC, where you can see for example just one or two frames from the animation, depending on the network speed. In some case, animations make the installation process even painful, see https://bugzilla.redhat.com/show_bug.cgi?id=1946064. No animation is good animation.

jstodola commented:
@AdamWill please be aware of this UI change, since it might affect your Fedora installation testing.

poncovka commented:
Also, add a release note for this change, please (https://anaconda-installer.readthedocs.io/en/latest/contributing.html#release-notes).

VladimirSlavik commented:
UI logic now works in absence of kickstart, onwards to that...

VladimirSlavik commented:
/kickstart-test --testtype smoke

VladimirSlavik commented:
/kcikstart-test --testtype users

VladimirSlavik commented:
/kickstart-test --testtype users

mairin commented:
@VladimirSlavik Hey! Taking a look now - 

Re: attribution - looks fine to me :)

Re: alignments... if the change you made to make the alignment intentionally different (e.g. the added indent on the description for each radio button option) prevents the shifting of elements you were worried about, the alignment screenshot you posted looks fine to me! I did examine the GNOME HIG and it does not account for the case where radio buttons have a header message + a non-header description, so this doesn't explicitly violate any design patterns outlined there, and I think it looks fine & clear.

Re: the animation - is there any way you could give me a preview so I could see what it looks like? Probably fade in / fade out will look fine, I just want to see how it comes off!

VladimirSlavik commented:
Thanks - and here you go!

https://user-images.githubusercontent.com/15903878/126683198-b5e887f1-f41e-450a-a551-6a883657eeb5.mp4

mairin commented:
@VladimirSlavik looks great to me, I see no issue!

jstodola commented:
The text boxes for entering the password are too wide - if they are as wide as possible, the screen would not look good on fullhd/4k monitors. Could the size of the password text boxes be reduced?
And one note about enabling/disabling remote access mentioned in the spoke - would it make sense to specify that it only affect remote access for the root account? Remote access for non-root users is not affected by this spoke.

VladimirSlavik commented:
Thanks for the reminder, I forgot about that part at some point. I changed the width to be 40 characters, not as wide as possible.

I also changed the wording - it's a separate commit so it can be squashed or dropped according to feelings about that.

![Screenshot_vs_main_2021-07-23_10:58:09](https://user-images.githubusercontent.com/15903878/126759986-f739383e-4d70-4832-999b-06f26c3d3d9f.png)

VladimirSlavik commented:
I don't think we are tracking any Fedora bugs / issues to link. The RHEL 9 ones are technically not related until this is ported. So linking nothing.

jstodola commented:
It looks great to me, thanks!

jkonecny12 commented on pyanaconda/ui/gui/spokes/root_password.py:
```diff
@@ -316,36 +316,32 @@ def _check_spoke_exit_conditions(self, unwaivable_check_failed):
-        self.password_entry.set_sensitive(not lock.get_active())
-        self.password_confirmation_entry.set_sensitive(not lock.get_active())
-        if not lock.get_active():
+    def on_disable_enable_clicked(self, control):
```
This method name seems pretty confusing to me. Right now I'm confused if it should toggle radio buttons or disable some and enable some. Could you please change the name?

VladimirSlavik replied:
Sorry, I don't agree. It's an event handler - the name conforms to the `on_controlname_eventtype` pattern most of our code uses. You do not call it, Gtk does. And the next line is a docstring, which also says what it is: _Click handler for enable and disable radio buttons._

jkonecny12 replied:
In that case is `disable_enable` a good control name?

VladimirSlavik replied:
The handler is used by two controls, one enables, one disables.

Maybe not mentioning the control names would make the name work? Eg. `on_root_enabled_changed`

jkonecny12 replied:
That seems much better to understand to me.

VladimirSlavik replied:
Great, I will do that, then.

VladimirSlavik replied:
Done.

jkonecny12 commented on pyanaconda/ui/gui/spokes/root_password.py:
```diff
@@ -79,8 +77,10 @@ def initialize(self):
         self._password_bar = self.builder.get_object("password_bar")
         self._password_label = self.builder.get_object("password_label")
-        self._lock = self.builder.get_object("lock")
+        self._enable_radio = self.builder.get_object("enable_radio")
+        self._disable_radio = self.builder.get_object("disable_radio")
```
I would like to change these variable names. I'm missing the target here. What to enable root account, ssh access, password field?

VladimirSlavik replied:
I added "root" in the middle of the names, eg. disable_root_radio.

jkonecny12 reviewed (changes requested):
Hi, I really like the new solution, however, I would like to make the variable names a bit more obvious.

poncovka commented:
> Thanks for the reminder, I forgot about that part at some point. I changed the width to be 40 characters, not as wide as possible.

I think you should limit the width of the whole action area to make it more centred.

VladimirSlavik commented:
Fun fact of the day: GTK3+ does not have a way to set maximum width. Instead, you have to make sure nothing within the widget of interest will expand. In this case, it was simple, the only expanding thing were the labels. As luck has it, labels, unlike other controls, do have a "maximum width" property... set in characters. Then, the "minimal width" can be set by adding a width request to something. I feel like it's 2000 again and I am trying to vertically center a thing on the web...

Here is 60 characters:

![Screenshot_vs_main_2021-07-29_22:02:06](https://user-images.githubusercontent.com/15903878/127558459-70da30a5-80b8-41ba-bcd3-97b140fe0daa.png)

VladimirSlavik commented:
Update: Limited width, rename radio buttons.

VladimirSlavik commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
I don't agree with the current `enable_disable` but I don't think it's reason for blocking this PR.

jkonecny12 reviewed (commented):
I don't agree with the current `enable_disable` but I don't think it's reason for blocking this PR.

VladimirSlavik commented:
Update: Changed the name of the handler per @jkonecny12 suggestion - thank you!

VladimirSlavik commented:
/kickstart-test --testtype smoke

jkonecny12 reviewed (approved):
Looks great to me now. Thanks!

VladimirSlavik commented:
/kickstart-test --testtype smoke

VladimirSlavik commented:
Update: Renamed to Root Account. The longer version with Security seems to still fit in English, but I wonder about all the translations.

poncovka reviewed (approved):
LGTM!

jkonecny12 reviewed (approved):
Still looks good to me!

VladimirSlavik commented:
@AdamWill this will break OpenQA, and #3538 too. I'm about to merge these so the breakage is imminent.

(While at it, I'd like to also get into next Fedora also some rehash of #2242. That would address the easier GUI wants. But that's only hopes so far.)
